ihmc-high-level-behaviors/src/libgdx/java/us/ihmc/rdx/ui/behavior/tree/RDXROS2BehaviorTree.java

```diff
@@ -19,49 +20,42 @@
  */
 public class RDXROS2BehaviorTree extends RDXBehaviorTree
 {
-   private final ROS2BehaviorTreeState ros2BehaviorTreeState;
+   private final ROS2BehaviorTree<RDXBehaviorTreeNode<?, ?>> ros2BehaviorTree;
    /** Reduce the communication update rate. */
-   private final Throttler communicationThrottler = new Throttler().setFrequency(ROS2BehaviorTreeState.SYNC_FREQUENCY);
+   private final Throttler communicationThrottler = new Throttler().setFrequency(ROS2BehaviorTree.SYNC_FREQUENCY);
    private final ImGuiAveragedFrequencyText subscriptionFrequencyText = new ImGuiAveragedFrequencyText();
    private final ImGuiAveragedFrequencyText publishFrequencyText = new ImGuiAveragedFrequencyText();
 
    public RDXROS2BehaviorTree(WorkspaceResourceDirectory treeFilesDirectory,
                               DRCRobotModel robotModel,
                               ROS2SyncedRobotModel syncedRobot,
+                              ROS2PeerClockOffsetEstimator peerClockEstimator,
                               RobotCollisionModel selectionCollisionModel,
                               RDXBaseUI baseUI,
                               RDX3DPanel panel3D,
                               ReferenceFrameLibrary referenceFrameLibrary,
                               ROS2ControllerPublishSubscribeAPI ros2)
    {
-      super(treeFilesDirectory,
-            robotModel,
-            syncedRobot,
-            selectionCollisionModel,
-            baseUI,
-            panel3D,
-            referenceFrameLibrary);
+      super(treeFilesDirectory, robotModel, syncedRobot, peerClockEstimator, selectionCollisionModel, baseUI, panel3D, referenceFrameLibrary);
 
-      ros2BehaviorTreeState = new ROS2BehaviorTreeState(getBehaviorTreeState(), this::setRootNode, ros2);
+      ros2BehaviorTree = new ROS2BehaviorTree<>(this, this::setRootNode, ros2);
 
-      ros2BehaviorTreeState.getBehaviorTreeSubscription().registerMessageReceivedCallback(subscriptionFrequencyText::ping);
+      ros2BehaviorTree.getBehaviorTreeSubscription().registerMessageReceivedCallback(subscriptionFrequencyText::ping);
    }
 
    public void update()
    {
-      boolean updateComms = communicationThrottler.run();
-      if (updateComms)
+      if (communicationThrottler.run())
       {
-         ros2BehaviorTreeState.updateSubscription();
+         // Must publish first to get newly modified data published
+         // This is because data gets modified after the update in the
+         // ImGui rendering thread.
+         ros2BehaviorTree.updatePublication();
+         publishFrequencyText.ping();
+         ros2BehaviorTree.updateSubscription();
```

Comment:
Important change, publication and subscription handling both happen before the update, which helps avoid edge cases and simplifies CRDT logic.